### PR TITLE
[WIP] Towards private/local Habitat Builder installations

### PR DESCRIPTION
**Pull request by schu**
**NB** not ready for merge yet.

This is an early work-in-progress pull request to provide a canonical, mostly automated Habitat Builder Vagrant setup as well as up-to-date and concise documentation on how to setup Habitat Builder in a private environment.

While some parts are still missing (most importantly good documentation on the integration with / configuration of the GitHub App) or to be done (e.g. get rid of the `/home/krangschnak` hack?), it shows

* what works for us today (after way too much trial-and-error) and
* that the process can (and IMO should) be both much simpler and better explained.

We would love to help with improvements to overcome the current situation, which is, fankly speaking, a jungle of setup scripts and documentation files :) Details on the problems can be found in e.g. https://github.com/habitat-sh/habitat/issues/4070 and https://github.com/habitat-sh/habitat/issues/4117

Feedback is very much welcome (and needed :).

**Our setup instructions (work-in-progress but working) can be found here**: https://github.com/kinvolk/habitat/blob/schu/vagrant/BUILDER_VAGRANT.md

Please note: while we have used Vagrant, I think the primary goal should be to have a canonical, automated and maintained setup and good documentation. If Docker is favored, we are happy to switch.

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: BUILDER_VAGRANT.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+Now, trigger a new build. For a package with no dependencies, above should
+be enough. Otherwise, repeat the process for every package reported
+missing during the build.
+
+If the build fails due to a missing public key, make sure you have both
+a public and a private key in the `/home/krangschnak/.hab/cache/keys/`
+directory, e.g.
+
+```
```
Comment (dongsupark) [inline BUILDER_VAGRANT.md]: How about adding `mkdir --parent /home/krangschnak/.hab/cache/keys`? In our test setup the directory is not there at all.

File: BUILDER_VAGRANT.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+cp /hab/cache/keys/foo-20171103084851.* /home/krangschnak/.hab/cache/keys/
+```
+
+TODO(schu): fix the weird key setup above ^
```
Comment (dongsupark) [inline BUILDER_VAGRANT.md]: +1 . This step looks always looks strange to me.

File: BUILDER_VAGRANT.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+## Troubleshooting
+
+* If you expirience authentication failures, make sure the direnv is
```
Comment (dongsupark) [inline BUILDER_VAGRANT.md]: s/expirience/experience/ ?

Comment (schu) [reply BUILDER_VAGRANT.md]: FIxed, thanks!

File: BUILDER_VAGRANT.md (hunk 1 above)
Comment (schu) [reply BUILDER_VAGRANT.md]: It's there for me, so there seems to be an inconsistency from the provisioning step depending on other factors. I have not yet found the code responsible.

Either way, I think we should rather get rid of this odd setup than add more steps to workaround.

**Comment by elliott-davis** [comment]
@schu are you intending this to run with airlock enabled or disabled?

**Comment by schu** [comment]
@elliott-davis sorry, late reply due to traveling.

> are you intending this to run with airlock enabled or disabled?

airlock is the habitat component responsible for chrooted studio environments? I don't have an opinion at the moment, since I'm not familiar with the inner workings of studio etc. (e.g. is airlock required for some parts?)

Ultimately I don't care much I think, as long as we can provide a convenient, automated build and setup. If airlock cannot run in a (privileged) container currently, that would have to be changed first for a Docker setup.

File: support/Procfile [hunk 4]
```diff
@@ -1,3 +1,4 @@
+web: support/builder_web.sh
```
Comment (elliott-davis) [inline support/Procfile]: This should probably be changed to `hab svc start core/builder-api-proxy`

File: Vagrantfile [hunk 5]
```diff
@@ -10,9 +10,12 @@ Vagrant.configure("2") do |config|
...
   config.vm.synced_folder "~/.hab/etc", "/hab/etc"
 
-  config.vm.network "forwarded_port", guest: 80, host: 9636
-  config.vm.network "forwarded_port", guest: 9631, host: 9631
-  config.vm.network "forwarded_port", guest: 9636, host: 9636
+  config.vm.network "private_network", ip: "192.168.198.7"
+
+  config.vm.network "forwarded_port", guest: 9631, host: 9631 # Supervisor API
+  config.vm.network "forwarded_port", guest: 9636, host: 9636 # Builder API
+  config.vm.network "forwarded_port", guest: 3000, host: 3000 # Builder UI
```
Comment (elliott-davis) [inline Vagrantfile]: You can change this to 80 as nobody as it's not really tenable to do web dev inside the vagrant vm (file syncing is super slow)

File: support/Procfile (hunk 4 above)
Comment (schu) [reply support/Procfile]: What is `builder-api-proxy` actually needed for? It seems to works just fine w/o (and is not mentioned in the new, authoritative [dev guide](https://github.com/habitat-sh/habitat/blob/master/BUILDER_DEV.md) nor in the [builder-web README](https://github.com/habitat-sh/habitat/blob/master/components/builder-web/README.md)).

File: Vagrantfile (hunk 5 above)
Comment (schu) [reply Vagrantfile]: Would it be ok to move the UI part into a Docker container instead?

**Comment by christophermaier** [comment]
@schu Is this still needed with the more recent changes around setting up a Builder locally?

**Comment by schu** [comment]
I guess it can be closed then. I don't work on/with Habitat currently but will give it a try when I get a chance. Thanks @christophermaier

**Pull request closed without merging**